include/deal.II/base/patterns_tools.h

```diff
@@ -0,0 +1,682 @@
+ * allow to convert from non elementary types to strings and viceversa.
+ *
+ * A typical usage of these tools is in the following example:
+ *
+ * @code
+ * typedef std::vector<unsigned int> T;
+ *
+ * T vec(3);
+ * vec[0] = 1;
+ * vec[1] = 3;
+ * vec[2] = 5;
+ *
+ * auto pattern = PatternsTools::Convert<T>::to_pattern();
+ *
+ * std::cout << pattern->description() << std::endl;
+ * // [List of <[Integer]> of length 0...4294967295 (inclusive)]
+ *
+ * auto s = PatternsTools::Convert<T>::to_string(vec);
+ * std::cout << s << std::endl;
+ * // 1, 2, 3
+ *
+ * auto vec = PatternsTools::Convert<T>::to_value("2,3,4,5");
+ * // now vec has size 4, and contains the elements 2,3,4,5
+ *
+ * std::cout << internal::RankInfo<T>::list_rank << std::endl; // Outputs 1
+ * std::cout << internal::RankInfo<T>::map_rank  << std::endl; // Outputs 0
+ * @endcode
+ *
+ * Convert<T> is used by the function PatternsTools::add_parameter() in this
+ * namespace. Internally it uses the internal::RankInfo<T> class to decide how
+ * many different separators are required to convert the given type to a string.
+ *
+ * For example, to write vectors of vectors, the default is to use "," for the first
+ * (inner) separator, and ";" for the second (outer) separator, i.e.
+ *
+ * @code
+ * std::vector<std::vector<unsigned int>> vec;
+ * vec = Convert<decltype(vec)>::to_value("1,2,3 ; 4,5,6");
+ *
+ * s = convert<decltype(vec[0])>::to_string(vec[0]);
+ * // s now contains the string "1,2,3"
+ * @endcode
+ *
+ * Separators for List and Map compatible types are selected according to the
+ * rank of the list and map objects, using the arrays
+ * PatternsTools::internal::default_list_separator and
+ * PatternsTools::internal::default_map_separator.
+ *
+ * They are currently set to:
+ *
+ * @code
+ * default_list_separator{{" ",  ","  ,  ";"  ,  "|"  ,   "%"}};
+ * default_map_separator {{" ",  ":"  ,  "="  ,  "@"  ,   "#"}};
+ * @endcode
+ *
+ * When one needs a mixture of Lists and Map types, their RankInfo is computed by taking
+ * the maximum of the vector_rank of the Key and of the Value type, so that, for example,
+ * it is possible to have the following
+ * @code
+ * ... // Build compare class
+ * std::map<std::vector<unsigned int>, std::vector<double>, compare> map;
+ *
+ * map = convert<decltype(map)>::to_value("1,2,3 : 5.0,6.0,7.0  ; 8,9,10 : 11.0,12.0,13.0");
+ *
+ * @endcode
+ *
+ * Some non elementary types are supported, like Point<dim>(), or std::complex<double>.
+ * If you wish to support more types, you have to specialize the Convert struct as well
+ * as the RankInfo struct.
+ *
+ * @ingroup input
+ * @author Luca Heltai, 2017
+ */
+namespace PatternsTools
+{
+  /**
+   * Converter class. This class is used to generate strings and Patterns associated to
+   * the given type, and to convert from a string to the given type and viceversa.
+   */
+  template<class T, class Enable=void>
+  struct Convert
```

Comment:
Yes, that occurred to me as well later on. I think this makes sense even though it looks a bit funny.

Maybe add something like "While the current function (in the general Convert template) is deleted, it is implemented and available in the specializations of the Convert class template for particular kinds of template arguments @p T."